linux/parser.bsh

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env bash
+
+if [[ $- != *i* ]]; then
+  source_once &> /dev/null && return 0
+fi
+
+#*# linux/parser
+
+#**
+# ======
+# Parser
+# ======
+#
+# .. default-domain:: bash
+#
+# .. file:: parser.bsh
+#
+# Functions to parse and/or validate common strings like URLs, etc.
+#**
+
+#**
+# .. function:: parse_url
+#
+# Parse a URL
+#
+# Parse a URL into its various parts: scheme, authority, user, host, port, rpath, query, fragment. Parsing URLs 100% correctly is HARD. This is not that solution.
+#
+# :Arguments: ``$1`` - URL
+#
+# .. rubric:: Example
+#
+# .. code-block:: bash
+#
+# parse_url https://visionsystemsinc.github.io:443/vsi_common/just/quickstart_user.html?highlight=J.U.S.T#what-is-j-u-s-t
+#
+# :Output:
+#   ``just_url_parsed_scheme`` - e.g., https
+#   ``just_url_parsed_authority`` - e.g., visionsystemsinc.github.io:443
+#   ``just_url_parsed_host`` - e.g., visionsystemsinc.github.io
+#   ``just_url_parsed_port`` - e.g., 443
+#   ``just_url_parsed_rpath`` - e.g., vsi_common/just/quickstart_user.html
+#   ``just_url_parsed_query`` - e.g., highlight=J.U.S.T
+#   ``just_url_parsed_fragment`` - e.g., what-is-j-u-s-t
+#
+# .. note::
+#    From https://stackoverflow.com/a/45977232
+#**
+
+function parse_url()
+{
+  arg="${1}"
+
+  local URI_BASIC_REGEX='^(([^:/?#]+):)?(//((([^:/?#]+)@)?([^:/?#]+)(:([0-9]+))?))?$'
+  # Note: No quotes around URI_REGEX
+  if [[ "${arg}" =~ ${URI_BASIC_REGEX} ]]; then
+    arg="${arg}/"
+  fi
+
+  # Following regex is based on https://tools.ietf.org/html/rfc3986#appendix-B
+  # with additional sub-expressions to split authority into userinfo, host and
+  # port
+  local URI_REGEX='^(([^:/?#]+):)?(//((([^:/?#]+)@)?([^:/?#]+)(:([0-9]+))?))?(/([^?#]*))(\?([^#]*))?(#(.*))?'
```

Comment:
Can we make this a global variable (set outside the function) so anyone who sources `parser.bsh` can use it?